[RISCV] Construct constants via instructions if materialization is costly

wangpc-pp opened the pull request:
For RISCV, it is costly to materialize constants used in lowering
`ISD::CTPOP`/`ISD::VP_CTPOP`.
    
We can query the materialization cost via `RISCVMatInt::getIntMatCost`
and if the cost is larger than 2, we should construct the constant
via two instructions.
    
This PR contains only vector part. We don't do the optimization if
LMUL is equal to LMUL_8 as it increases register preasure.

wangpc-pp commented:
There is a problem that we will lowering `ISD::CTLZ/ISD::CTTZ` to `ISD::CTPOP` if `ISD::CTPOP` is `Legal` or `Custom`, which causes some regressions. I am still fixing it.

efriedma-quic commented:
For the vector case, I think it ends up cheaper if you always start by bitcasting the operand to an i8 vector, and doing an i8 popcount.  That makes the constants involved significantly cheaper. (The width you use to do the arithmetic doesn't matter for correctness, but doing it in i8 means you need fewer vsetvli, I think).  Then you can bitcast back to the wider type to do the multiply/shift.

wangpc-pp commented:
> For the vector case, I think it ends up cheaper if you always start by bitcasting the operand to an i8 vector, and doing an i8 popcount. That makes the constants involved significantly cheaper. (The width you use to do the arithmetic doesn't matter for correctness, but doing it in i8 means you need fewer vsetvli, I think). Then you can bitcast back to the wider type to do the multiply/shift.

I don't know if this is benefical. Using smaller EEW means that we can simplify the materialization but we may execute more uops I think, which may result in worse performance.

topperc commented:
> > For the vector case, I think it ends up cheaper if you always start by bitcasting the operand to an i8 vector, and doing an i8 popcount. That makes the constants involved significantly cheaper. (The width you use to do the arithmetic doesn't matter for correctness, but doing it in i8 means you need fewer vsetvli, I think). Then you can bitcast back to the wider type to do the multiply/shift.
> 
> I don't know if this is benefical. Using smaller EEW means that we can simplify the materialization but we may execute more uops I think, which may result in worse performance.

Hopefully simple bitwise operations like shifts, ands, and adds don't get split based on element width.

wangpc-pp commented:
> > > For the vector case, I think it ends up cheaper if you always start by bitcasting the operand to an i8 vector, and doing an i8 popcount. That makes the constants involved significantly cheaper. (The width you use to do the arithmetic doesn't matter for correctness, but doing it in i8 means you need fewer vsetvli, I think). Then you can bitcast back to the wider type to do the multiply/shift.
> > 
> > 
> > I don't know if this is benefical. Using smaller EEW means that we can simplify the materialization but we may execute more uops I think, which may result in worse performance.
> 
> Hopefully simple bitwise operations like shifts, ands, and adds don't get split based on element width.

Yeah, most instructions will be split to chunks of datapath length in most implementations, but I do know an implementation that doesn't.

wangpc-pp committed:
[RISCV] Construct constants via instructions if materialization is costly

For RISCV, it is costly to materialize constants used in lowering
`ISD::CTPOP`/`ISD::VP_CTPOP`.

We can query the materialization cost via `RISCVMatInt::getIntMatCost`
and if the cost is larger than 2, we should construct the constant
via two instructions.

This PR contains only vector part. We don't do the optimization if
LMUL is equal to LMUL_8 as it increases register preasure.

wangpc-pp commented:
Only vector part is included and we don't optimize if LMUL==M8 as it increases register preasure. Now there is no noticeable regression.

topperc commented:
> > > > For the vector case, I think it ends up cheaper if you always start by bitcasting the operand to an i8 vector, and doing an i8 popcount. That makes the constants involved significantly cheaper. (The width you use to do the arithmetic doesn't matter for correctness, but doing it in i8 means you need fewer vsetvli, I think). Then you can bitcast back to the wider type to do the multiply/shift.
> > > 
> > > 
> > > I don't know if this is benefical. Using smaller EEW means that we can simplify the materialization but we may execute more uops I think, which may result in worse performance.
> > 
> > 
> > Hopefully simple bitwise operations like shifts, ands, and adds don't get split based on element width.
> 
> Yeah, most instructions will be split to chunks of datapath length in most implementations, but I do know an implementation that doesn't.

Can you share what implementation that is?

wangpc-pp commented:
> > > > > For the vector case, I think it ends up cheaper if you always start by bitcasting the operand to an i8 vector, and doing an i8 popcount. That makes the constants involved significantly cheaper. (The width you use to do the arithmetic doesn't matter for correctness, but doing it in i8 means you need fewer vsetvli, I think). Then you can bitcast back to the wider type to do the multiply/shift.
> > > > 
> > > > 
> > > > I don't know if this is benefical. Using smaller EEW means that we can simplify the materialization but we may execute more uops I think, which may result in worse performance.
> > > 
> > > 
> > > Hopefully simple bitwise operations like shifts, ands, and adds don't get split based on element width.
> > 
> > 
> > Yeah, most instructions will be split to chunks of datapath length in most implementations, but I do know an implementation that doesn't.
> 
> Can you share what implementation that is?

Just checked, I was wrong as the implementation doesn't split shift/and/add based on EEW (It's a small core targeting for low-end IoT scenarios and vlen is really small). Sorry for that.
Besides, maybe we can do some optimizations for more possible operations if this `simple arithmetics don't get split based on EEW` assumption is right?
